### PR TITLE
Category

### PR DESCRIPTION
**Pull request by Dima23-ops5**

### CODE REVIEW

File: src/main/java/mate/academy/springbootintro/controller/CategoryController.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            description = "Delete category by id"
+    )
+    public void deleteCategory(@PathVariable @Positive Long id) {
+        categoryService.deleteById(id);
+    }
+
+    @PreAuthorize("hasRole('USER')")
+    @GetMapping("/{id}/books")
+    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(Long id) {
+        return null;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/controller/CategoryController.java]: implement this endpoint

File: src/main/java/mate/academy/springbootintro/dto/bookdto/CreateBookRequestDto.java [hunk 2]
```diff
@@ -19,4 +21,6 @@ public class CreateBookRequestDto {
     private String description;
     @NotBlank(message = "CoverImage can`t be null")
     private String coverImage;
+    @NotBlank
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/dto/bookdto/CreateBookRequestDto.java]:
```suggestion
    @NotEmpty
```

for Collection

File: src/main/java/mate/academy/springbootintro/dto/bookdto/CreateBookRequestDto.java [hunk 3]
```diff
@@ -19,4 +21,6 @@ public class CreateBookRequestDto {
     private String description;
     @NotBlank(message = "CoverImage can`t be null")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/dto/bookdto/CreateBookRequestDto.java]:
```suggestion
    @NotBlank(message = "Cover image can`t be null")
```

File: src/main/java/mate/academy/springbootintro/repository/book/filter/BookSpecificationBuilder.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+import mate.academy.springbootintro.dto.bookdto.BookSearchParametersDto;
+import mate.academy.springbootintro.model.Book;
+import mate.academy.springbootintro.repository.SpecificationBuilder;
+import org.springframework.data.jpa.domain.Specification;
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class BookSpecificationBuilder implements SpecificationBuilder<Book> {
+    private static final String Author = "author";
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/repository/book/filter/BookSpecificationBuilder.java]:
```suggestion
    private static final String AUTHOR = "author";
```

File: src/main/java/mate/academy/springbootintro/security/JwtAuthenticationFilter.java [hunk 5]
```diff
@@ -4,23 +4,25 @@
...
 import org.springframework.web.filter.OncePerRequestFilter;
 
-import java.io.IOException;
-
 @Component
 @RequiredArgsConstructor
 public class JwtAuthenticationFilter extends OncePerRequestFilter {
-
+    private static final String AUTHORIZATION = "Authorization";
+    private static final String BEARER = "Bearer";
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/security/JwtAuthenticationFilter.java]:
```suggestion
    private static final String TOKEN_HEADER = "Bearer ";
```

File: src/main/java/mate/academy/springbootintro/security/JwtAuthenticationFilter.java [hunk 6]
```diff
@@ -39,13 +41,13 @@ protected void doFilterInternal(
...
+        filterChain.doFilter(request, response);
     }
 
     private String getToken(HttpServletRequest request) {
-        String bearerToken = request.getHeader("Authorization");
-        if (bearerToken != null && bearerToken.startsWith("Bearer")) {
-            return bearerToken.substring(7);
+        String bearerToken = request.getHeader(AUTHORIZATION);
+        if (bearerToken != null && bearerToken.startsWith(BEARER)) {
+            return bearerToken.substring(DEFAULT_CAPACITY);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/security/JwtAuthenticationFilter.java]:
```suggestion
            return bearerToken.substring(TOKEN_HEADER.length());
```

File: src/main/java/mate/academy/springbootintro/security/JwtAuthenticationFilter.java [hunk 7]
```diff
@@ -39,13 +41,13 @@ protected void doFilterInternal(
...
         }
-       filterChain.doFilter(request, response);
+        filterChain.doFilter(request, response);
     }
 
     private String getToken(HttpServletRequest request) {
-        String bearerToken = request.getHeader("Authorization");
-        if (bearerToken != null && bearerToken.startsWith("Bearer")) {
-            return bearerToken.substring(7);
+        String bearerToken = request.getHeader(AUTHORIZATION);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/security/JwtAuthenticationFilter.java]:
```suggestion
        String bearerToken = request.getHeader(HttpHeaders.AUTHORIZATION);
```

File: src/main/java/mate/academy/springbootintro/security/JwtAuthenticationFilter.java [hunk 8]
```diff
@@ -4,23 +4,25 @@
...
 
-import java.io.IOException;
-
 @Component
 @RequiredArgsConstructor
 public class JwtAuthenticationFilter extends OncePerRequestFilter {
-
+    private static final String AUTHORIZATION = "Authorization";
+    private static final String BEARER = "Bearer";
+    private static final int DEFAULT_CAPACITY = 7;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/security/JwtAuthenticationFilter.java]:
```suggestion
```

File: src/main/java/mate/academy/springbootintro/service/book/impl/BookServiceImpl.java [hunk 9]
```diff
@@ -64,4 +65,11 @@ public List<BookDto> search(BookSearchParametersDto searchParameters) {
                 .toList();
     }
 
+    @Override
+    public List<BookDtoWithoutCategoryIds> getBooksByCategoryId(Long id) {
+        return bookRepository.getBooksByCategoryId(id).stream()
+                .map(bookMapper::toDtoWithoutCategories)
+                .toList();
+    }
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/springbootintro/service/book/impl/BookServiceImpl.java]:
```suggestion
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

**Review by sarakhmen — APPROVED** [review]

**Pull request closed without merging**